### PR TITLE
Personalization layers on MPI and MPI Serial

### PR DESCRIPTION
**Pull request by shourya01**
Code for personalization layers, applied to federated load forecasting on NREL ComStock dataset. Currently only implemented for MPI.

### CODE REVIEW

File: examples/losses/maeloss.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import torch.nn as nn
+
+class MAELoss(nn.Module):
+    '''Mean Squared Error Loss'''
```
Comment (Zilinghan) [inline examples/losses/maeloss.py:4]: I guess the comment here should be Mean Absolute Error Loss?

File: examples/metric/mae.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import numpy as np
+
+def ae(y_true, y_pred):
```
Comment (Zilinghan) [inline examples/metric/mae.py:3]: Maybe rename this to mae (Mean Absolute Error)

File: examples/losses/maeloss.py (hunk 1 above)
Comment (shourya01) [reply examples/losses/maeloss.py:4]: right, my bad

File: examples/metric/mae.py (hunk 2 above)
Comment (shourya01) [reply examples/metric/mae.py:3]: sure. I added the .mean() and updated the filename but for some reason forgot to update function name

File: examples/personalization_fedloadforecast.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+import appfl.run_mpi as rm
+
+from models.cnn import *
+from models.utils import get_model
+from models.utils import validate_parameter_names
+
+from losses.utils import get_loss
+from metric.utils import get_metric
+
+from datasets.PreprocessedData.NREL_Preprocess import NRELDataDownloader
```
Comment (Zilinghan) [inline examples/personalization_fedloadforecast.py:23]: Could you please also include code for NRELDataDownloader

File: src/appfl/algorithm/client_optimizer.py
Comment (Zilinghan) [inline src/appfl/algorithm/client_optimizer.py:1]: For this file, I suggest creating another client optimizer called personalized_fl_optimizer.py

**Review by Zilinghan — CHANGES_REQUESTED** [review]
Please see the individual comments, thanks

File: examples/personalization_fedloadforecast.py (hunk 3 above)
Comment (shourya01) [reply examples/personalization_fedloadforecast.py:23]: Ah I just saw this. Looks like PreprocessedData is blacklisted in .gitignore. will add

**Review by kibaekkim — COMMENTED** [review]
@shourya01 Can you add some descriptions in `src/appfl/run_serial.py` and `src/appfl/rum_mpi.py`? This would make others quickly read the code.

@Zilinghan I wonder if the files/functions in `examples/losses` and `examples/metric` would be reusable later. They look useful. If so, we can think about that when refactoring the package.

**Comment by Zilinghan** [comment]
> @shourya01 Can you add some descriptions in `src/appfl/run_serial.py` and `src/appfl/rum_mpi.py`? This would make others quickly read the code.
> 
> @Zilinghan I wonder if the files/functions in `examples/losses` and `examples/metric` would be reusable later. They look useful. If so, we can think about that when refactoring the package.

Yes, they are definitely reusable, now  `examples/losses` and `examples/metric` only have cross entropy loss and percentage accuracy.

File: src/appfl/misc/utils.py [hunk 4]
```diff
@@ -121,6 +123,55 @@ def save_model_iteration(t, model, cfg: DictConfig):
...
+        file = cfg.load_model_dirname + "/%s%s" % (cfg.load_model_filename, ".pt")
+    else:
+        file = cfg.load_model_dirname + "/client_%d"%client_id + "/%s%s" % (cfg.load_model_filename, ".pt")
+
+    model.load_state_dict(torch.load(file),strict=False)
+        
+    return model
+    
+
+def save_model_state_iteration(t, model, cfg: DictConfig, client_id = None):
```
Comment (kibaekkim) [inline src/appfl/misc/utils.py:142]: Is this used for `personalized FL`? Can we combine this with `save_model_iteration`?

Comment (shourya01) [reply src/appfl/misc/utils.py:142]: Yes, it is used for `personalized FL`.

It will be a bit difficult to combine it with `save_model_iteration`. `save_model_iteration` currently saves the whole model, while `save_model_state_iteration` saves the `state_dict` of the model, and excludes the personalized weights from the `state_dict` if we are saving the server weights.

Comment (shourya01) [reply src/appfl/misc/utils.py:142]: One way to merge it with `save_model_iteration` would be to zero out the personalized weights in the server model before saving it.

**Review by kibaekkim — APPROVED** [review]
